pkg/controller/volume/cache/actual_state_of_world.go

```diff
@@ -0,0 +1,314 @@
+type ActualStateOfWorld interface {
+	// AddVolumeNode adds the given volume and node to the underlying store
+	// indicating the specified volume is attached to the specified node.
+	// A unique volumeName is generated from the volumeSpec and returned on
+	// success.
+	// If the volume/node combo already exists, this is a no-op.
+	// If volumeSpec is not an attachable volume plugin, an error is returned.
+	// If no volume with the name volumeName exists in the store, the volume is
+	// added.
+	// If no node with the name nodeName exists in list of attached nodes for
+	// the specified volume, the node is added.
+	AddVolumeNode(volumeSpec *volume.Spec, nodeName string) (string, error)
+
+	// MarkVolumeNodeSafeToDetach marks the specified volume/node combo as safe
+	// to detach.
+	// If no volume with the name volumeName exists in the store, an error is
+	// returned.
+	// If no node with the name nodeName exists in list of attached nodes for
+	// the specified volume, an error is returned.
+	MarkVolumeNodeSafeToDetach(volumeName, nodeName string) error
+
+	// MarkDesireToDetach returns the difference between the current time  and
+	// the DetachRequestedTime for the given volume/node combo. If the
+	// DetachRequestedTime is zero, it is set to the current time.
+	// If no volume with the name volumeName exists in the store, an error is
+	// returned.
+	// If no node with the name nodeName exists in list of attached nodes for
+	// the specified volume, an error is returned.
+	MarkDesireToDetach(volumeName, nodeName string) (time.Duration, error)
+
+	// DeleteVolumeNode removes the given volume and node from the underlying
+	// store indicating the specified volume is no longer attached to the
+	// specified node.
+	// If the volume/node combo does not exist, this is a no-op.
+	// If after deleting the node, the specified volume contains no other child
+	// nodes, the volume is also deleted.
+	DeleteVolumeNode(volumeName, nodeName string)
+
+	// VolumeNodeExists returns true if the specified volume/node combo exists
+	// in the underlying store indicating the specified volume is attached to
+	// the specified node.
+	VolumeNodeExists(volumeName, nodeName string) bool
+
+	// GetAttachedVolumes generates and returns a list of volumes/node pairs
+	// reflecting which volumes are attached to which nodes based on the
+	// current actual state of the world.
+	GetAttachedVolumes() []AttachedVolume
+}
+
+// AttachedVolume represents a volume that is attached to a node.
+type AttachedVolume struct {
+	// VolumeName is the unique identifier for the volume that is attached.
+	VolumeName string
+
+	// VolumeSpec is the volume spec containing the specification for the
+	// volume that is attached.
+	VolumeSpec *volume.Spec
+
+	// NodeName is the identifier for the node that the volume is attached to.
+	NodeName string
+
+	// SafeToDetach indicates that this node/volume combo has been unmounted
```

Comment:
maybe reword to "indicates that this volume has been unmount from the node and is..."